spring-cloud-dataflow-admin-local/src/main/java/org/springframework/cloud/dataflow/admin/spi/local/OutOfProcessModuleDeployer.java

```diff
@@ -0,0 +1,233 @@
+public class OutOfProcessModuleDeployer implements ModuleDeployer {
+
+	private String moduleLauncherPath;
+
+	private static final Logger logger = LoggerFactory.getLogger(OutOfProcessModuleDeployer.class);
+
+	private Map<ModuleDeploymentId, List<Instance>> running = new HashMap<>();
+
+	@Autowired
+	private OutOfProcessModuleDeployerProperties properties;
+
+	@Override
+	public ModuleDeploymentId deploy(ModuleDeploymentRequest request) {
+
+		String module = request.getCoordinates().toString();
+
+		ModuleDeploymentId moduleDeploymentId = ModuleDeploymentId.fromModuleDefinition(request.getDefinition());
+		List<Instance> processes = new ArrayList<>();
+		running.put(moduleDeploymentId, processes);
+
+
+		HashMap<String, String> args = new HashMap<>();
+		args.put("modules", request.getCoordinates().toString());
+		args.putAll(ModuleArgumentQualifier.qualifyArgs(0, request.getDefinition().getParameters()));
+		args.putAll(ModuleArgumentQualifier.qualifyArgs(0, request.getDeploymentProperties()));
+		String jmxDomainName = String.format("%s.%s", request.getDefinition().getGroup(), request.getDefinition().getLabel());
+		args.putAll(ModuleArgumentQualifier.qualifyArgs(0, Collections.singletonMap(JMX_DEFAULT_DOMAIN_KEY, jmxDomainName)));
+
+		args.put("endpoints.shutdown.enabled", "true");
+		args.put("endpoints.jmx.unique-names", "true");
+
+
+		for (int i = 0; i < request.getCount(); i++) {
+			int port = SocketUtils.findAvailableTcpPort(DEFAULT_SERVER_PORT);
+			args.put(SERVER_PORT_KEY, String.valueOf(port));
+
+			ProcessBuilder builder = new ProcessBuilder(properties.getJavaCmd(), "-jar", moduleLauncherPath);
+			builder.environment().clear();
+			builder.environment().putAll(args);
+
+			try {
+				Path workDir = Files.createTempDirectory(properties.getWorkingDirectoriesRoot(), moduleDeploymentId.toString() + "-" + i + "-");
+				if (properties.isDeleteFilesOnExit()) {
+					workDir.toFile().deleteOnExit();
+				}
+
+				processes.add(new Instance(moduleDeploymentId, i, builder, workDir));
+				logger.info("deploying module {} instance {}\n   Logs will be in {}", module, i, workDir);
+			}
+			catch (IOException e) {
+				throw new RuntimeException("Exception trying to deploy " + request, e);
+			}
+
+		}
+
+
+		return moduleDeploymentId;
+	}
+
+	@Override
+	public void undeploy(ModuleDeploymentId id) {
+		List<Instance> processes = running.get(id);
+		if (processes != null) {
+			RestTemplate restTemplate = new RestTemplate();
+			for (Instance instance : processes) {
+				if (!isAlive(instance.process)) {
+					continue;
+				}
+				try {
+					restTemplate.postForObject(instance.moduleUrl + "/shutdown", null, String.class);
+					instance.process.waitFor();
+				}
+				catch (InterruptedException e) {
+					instance.process.destroy();
+				}
+			}
+			running.remove(id);
+		}
+	}
+
+	@Override
+	public ModuleStatus status(ModuleDeploymentId id) {
+		List<Instance> instances = running.get(id);
+		if (instances == null) {
+			return null;

```

Comment:
We shouldn't return `null` here, probably `unknown` until https://github.com/spring-cloud/spring-cloud-dataflow/pull/194 is merged, when it would be `undeployed`. In particular, this results in an NPE for streams that are created but not deployed.